Userland/Libraries/LibSemVer/SemVer.cpp

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) 2023, Gurkirat Singh <tbhaxor@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/CharacterTypes.h>
+#include <AK/Find.h>
+#include <AK/GenericLexer.h>
+#include <AK/ReverseIterator.h>
+#include <AK/StringBuilder.h>
+#include <LibSemVer/SemVer.h>
+
+namespace SemVer {
+String SemVer::suffix() const
+{
+    StringBuilder sb;
+    if (!m_prerelease_identifiers.is_empty())
+        sb.appendff("-{}", prerelease());
+    if (!m_buildmetadata.is_empty())
+        sb.appendff("+{}", buildmetadata());
+    return sb.to_string().release_value_but_fixme_should_propagate_errors();
+}
+
+String SemVer::to_string() const
+{
+    return String::formatted("{}{}{}{}{}{}", m_major, m_number_separator, m_minor, m_number_separator, m_patch, suffix()).release_value_but_fixme_should_propagate_errors();
+}
+
+SemVer SemVer::bump(BumpType type) const
+{
+    switch (type) {
+    case BumpType::Major:
+        return SemVer(m_major + 1, 0, 0, m_number_separator);
+    case BumpType::Minor:
+        return SemVer(m_major, m_minor + 1, 0, m_number_separator);
+    case BumpType::Patch:
+        return SemVer(m_major, m_minor, m_patch + 1, m_number_separator);
+    case BumpType::Prerelease: {
+        Vector<String> prerelease_identifiers = m_prerelease_identifiers;
+        bool is_found = false;
+
+        // Unlike comparision, prerelease bumps take from RTL.
+        for (auto& identifier : AK::ReverseWrapper::in_reverse(prerelease_identifiers)) {
+            auto numeric_identifier = identifier.to_number<u32>();
+            if (numeric_identifier.has_value()) {
+                is_found = true;
+                identifier = String::formatted("{}", numeric_identifier.value() + 1).release_value_but_fixme_should_propagate_errors();
+                break;
+            }
+        }
+
+        // Append 0 identifier if there is no numeric found to be bumped.
+        if (!is_found)
+            prerelease_identifiers.append("0"_string);
+
+        return SemVer(m_major, m_minor, m_patch, m_number_separator, prerelease_identifiers, ""_string);
+    }
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+bool SemVer::is_same(SemVer const& other, CompareType compare_type) const
+{
+    switch (compare_type) {
+    case CompareType::Major:
+        return m_major == other.m_major;
+    case CompareType::Minor:
+        return m_major == other.m_major && m_minor == other.m_minor;
+    case CompareType::Patch:
+        return m_major == other.m_major && m_minor == other.m_minor && m_patch == other.m_patch;
+    default:
+        return to_string() == other.to_string();
```

Comment:
Do you have any rationale for this? To me exact match means "_literally_" exact match.

Here is my thought: What if the build metadata contains certain build information which is used to debug the bugs.

Reply:
Sure, the SemVer spec item 10:

https://semver.org/#spec-item-10

> Build metadata MUST be ignored when determining version precedence. **Thus two versions that differ only in the build metadata, have the same precedence.** Examples: 1.0.0-alpha+001, 1.0.0+20130313144700, 1.0.0-beta+exp.sha.5114f85, 1.0.0+21AF26D3----117B344092BD.

Reply:
Ah, misunderstanding again. This makes sense to me.

Reply:
Now I discovered, it also has identifier list as like pre-release. Vector<String> instead of String. Thanks for pointing out the bug, this will also be fixed along with the one you pointed out.

Reply:
I have still confusion on the bolded line, but lets not include build metadata for now.